src/kernel/capinv.c

```diff
@@ -877,6 +878,7 @@ composite_syscall_handler(struct pt_regs *regs)
 	 * Thus the level of indirection here.
 	 */
 	if (!thd_switch) __userregs_set(regs, ret, __userregs_getsp(regs), __userregs_getip(regs));
+	else return ret; /* ret from cap_switch */
```

Comment:
Shouldn't this logic be:

```
if (thd_switch) return ret;
__userregs_set(regs, ret,  __userregs_getsp(regs), __userregs_getip(regs));
return 0;
```

???

The current conditional structure seems overly complicated and non-uniform (one branch returning, the other falling through, but with no textual indication that it is falling through.